### PR TITLE
Remove custom implementation of MurmurHash2 with browser implementation of sha-256

### PR DESCRIPTION
**Pull request by JakeChampion**
In the actual JavaScript code -- 93 changes: 22 additions & 71 deletions

:)

### CODE REVIEW

**Comment by origamiserviceuser** [comment]
o-tracking bundle size difference from 2.0.10 to b6c78d2ff5e2c278525e57e8d09047400d7f20d4
js: 12.88kb increase (4.25kb/gzip)

File: MIGRATION.md [hunk 1]
```diff
@@ -5,6 +5,8 @@
 
 o-tracking now uses the Beacon API in all browsers which support it. In v2 this was an opt-in feature, now it is always enabled in browsers which support it. For browsers which do not support Beacon API o-tracking will revert to the XMLHTTPRequest API.
 
+o-tracking now uses the Crypto, TextEncoder, and Uint8Array APIs, to support older browser please include polyfills for these features.
```
Comment (notlee) [inline MIGRATION.md:8]: Does polyfill.io cover all of these? A link or some more instruction on how would be helpful for people migrating :) especially for Build Service users, I think it's common that o-tracking is included with the Build Service -- might even be worth flagging in the readme if needed for IE11?

File: test/events/custom.test.js [hunk 2]
```diff
@@ -50,7 +50,7 @@ describe('event', function () {
 		proclaim.equal(sent_data.context.component_name, 'custom-o-tracking');
 	});
 
-	it('should listen to a dom event and generate a component_id', function () {
+	it.only('should listen to a dom event and generate a component_id', function (done) {
```
Comment (notlee) [inline test/events/custom.test.js:53]:
```suggestion
	it('should listen to a dom event and generate a component_id', function (done) {
```

**Review by notlee — COMMENTED** [review]
this is nice! think some more docs might be useful to help users migrate though -- if I need to support IE11 and use the Build Service, can I include everything I need from polyfill.io?

File: test/events/custom.test.js (hunk 2 above)
Comment (JakeChampion) [reply test/events/custom.test.js:53]: 🤦

File: MIGRATION.md (hunk 1 above)
Comment (JakeChampion) [reply MIGRATION.md:8]: polyfill.io does not have Crypto.subtle but does have the rest of them

**Comment by origamiserviceuser** [comment]
o-tracking bundle size difference from 2.0.10 to 941cea24bf11cc624c29d1cfec373c0647599bea
js: 12.88kb increase (4.25kb/gzip)

File: MIGRATION.md (hunk 1 above)
Comment (JakeChampion) [reply MIGRATION.md:8]: I'm running the tests on browserstack across all our supported browsers to confirm it works

**Comment by JakeChampion** [comment]
Looks to fail using obt browserstack but when I test those browsers myself they seemed to pass the tests...
```
Failed Karma tests: 

      Chrome 83.0.4103.83 (Android 5.0) failed specs:
      event > should listen to a dom event and generate a component_id
        TypeError: Cannot read property 'digest' of undefined
            at getComponentId (src/javascript/events/custom.js:134:1 <- test/events/custom.test.js:1448:37)
            at event (src/javascript/events/custom.js:57:1 <- test/events/custom.test.js:1371:70)
            at HTMLBodyElement.<anonymous> (test/events/custom.test.js:71:1 <- test/events/custom.test.js:2182:11)
            at Context.<anonymous> (test/events/custom.test.js:91:1 <- test/events/custom.test.js:2202:17)
  
      Chrome 83.0.4103.83 (Android 5.0) failed specs:
      event > should listen to a dom event and generate a component_id
        TypeError: Cannot read property 'digest' of undefined
            at getComponentId (src/javascript/events/custom.js:134:1 <- test/events/custom.test.js:1448:37)
            at event (src/javascript/events/custom.js:57:1 <- test/events/custom.test.js:1371:70)
            at HTMLBodyElement.<anonymous> (test/events/custom.test.js:71:1 <- test/events/custom.test.js:2182:11)
            at Context.<anonymous> (test/events/custom.test.js:91:1 <- test/events/custom.test.js:2202:17)
  
      Mobile Safari 10.0 (iOS 10.3.3) failed specs:
      Core.Send > fallback transports > use sendBeacon by default
        Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.
    
      event > should listen to a dom event and generate a component_id
        undefined is not an object (evaluating 'crypto.subtle.digest')
        src/javascript/events/custom.js:134:1 <- test/events/custom.test.js:1448:36
        asyncFunctionResume@[native code]
        src/javascript/events/custom.js:57:1 <- test/events/custom.test.js:1371:84
        asyncFunctionResume@[native code]
        test/events/custom.test.js:71:1 <- test/events/custom.test.js:2182:21
        asyncFunctionResume@[native code]
        dispatchEvent@[native code]
        test/events/custom.test.js:91:1 <- test/events/custom.test.js:2202:30
  
      event > should throw custom error message if the tracking event object contains circular references
        o-tracking does not support circular references in the analytics data.
        Please remove the circular references in the data.
        Here are the paths in the data which are circular:
        [
            "[0].item.context.context.circular"
        ] == o-tracking does not support circular references in the analytics data.
        Please remove the circular references in the data.
        Here are the paths in the data which are circular:
        [
            ".context.context.circular"
        ]
  
      Firefox 80.0 (Mac OS 10.14) failed specs:
      event > should listen to a dom event and generate a component_id
        crypto.subtle is undefined
        getComponentId@src/javascript/events/custom.js:134:1 <- test/events/custom.test.js:1448:17
        event@src/javascript/events/custom.js:57:1 <- test/events/custom.test.js:1371:70
        file_custom$2etest$2ejs/</</</<@test/events/custom.test.js:71:1 <- test/events/custom.test.js:2182:11
        file_custom$2etest$2ejs/</</<@test/events/custom.test.js:91:1 <- test/events/custom.test.js:2202:17
  
  
      Chrome 85.0.4183.83 (Mac OS 10.14.6) failed specs:
      event > should listen to a dom event and generate a component_id
        TypeError: Cannot read property 'digest' of undefined
            at getComponentId (src/javascript/events/custom.js:134:1 <- test/events/custom.test.js:1448:37)
            at event (src/javascript/events/custom.js:57:1 <- test/events/custom.test.js:1371:70)
            at HTMLBodyElement.<anonymous> (test/events/custom.test.js:71:1 <- test/events/custom.test.js:2182:11)
            at Context.<anonymous> (test/events/custom.test.js:91:1 <- test/events/custom.test.js:2202:17)
  
      Edge 85.0.564.41 (Windows 10) failed specs:
      event > should listen to a dom event and generate a component_id
        TypeError: Cannot read property 'digest' of undefined
            at getComponentId (src/javascript/events/custom.js:134:1 <- test/events/custom.test.js:1448:37)
            at event (src/javascript/events/custom.js:57:1 <- test/events/custom.test.js:1371:70)
            at HTMLBodyElement.<anonymous> (test/events/custom.test.js:71:1 <- test/events/custom.test.js:2182:11)
            at Context.<anonymous> (test/events/custom.test.js:91:1 <- test/events/custom.test.js:2202:17)
  
      8 tests failed across 6 browsers.
     → Failed Karma tests
```

**Comment by JakeChampion** [comment]
Ohhhh it is because crypto.subtle is only available over secure contexts (https) and our tests run over http.

**Comment by origamiserviceuser** [comment]
o-tracking bundle size difference from 2.0.10 to 38cf8853b0c4ceaf6278081f091a78fa9ab89d6f
js: 12.91kb increase (4.26kb/gzip)

**Comment by notlee** [comment]
Couple more questions I just thought of:

- could the hash change based on browser implmentaiton e.g. `webkitSubtle` vs `subtle`?
- will the generated hash be different from the v2 implementation?

And if so is that a problem for the data team?

**Comment by JakeChampion** [comment]
>could the hash change based on browser implmentaiton e.g. `webkitSubtle` vs `subtle`?

No, it would be the same hash value because they use the same algorithm to generate the hash

>will the generated hash be different from the v2 implementation? And if so is that a problem for the data team?

Yes, browsers don't have a MurMurHash2 implemenation, so it would be different values generated.
The question about whether it's a problem for the data team will likely have to be answered by @chrisbrownuk as I don't know the answer.

**Comment by chrisbrownuk** [comment]
As far as I can see this is used to generate the context.component_id value. This is to allow "grouping for all events from the same component", so as long as a browser generates a self consistent id for a component it will be fine.

That being said I didn't recall ever seeing context.component_id.  Looking at Redshift I see that there are a tiny number of events with this.  They are either share|click or tabs|click events.  I don't know what actions generate these events.  Pretty sure no one uses these events for analysis.

**Comment by JakeChampion** [comment]
@chrisbrownuk it sounds like we can remove this piece of code entirely then! Is that correct?

**Comment by chrisbrownuk** [comment]
@JakeChampion I'd better just ask in a few places first, just to be safe.

**Comment by chrisbrownuk** [comment]
@JakeChampion I've had one vote to keep this functionality as it is useful if a custom name wasn't specified.

**Comment by JakeChampion** [comment]
Thank you for finding that out @chrisbrownuk 🙌

**Comment by JakeChampion** [comment]
> @JakeChampion I've had one vote to keep this functionality as it is useful if a custom name wasn't specified.

If we made supplying a custom name mandatory, would they be happy with this generated ID being removed?

**Pull request closed without merging**